gtdynamics/dynamics/ChainDynamicsGraph.cpp

```diff
@@ -0,0 +1,223 @@
+/* ----------------------------------------------------------------------------
+ * GTDynamics Copyright 2020, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * See LICENSE for the license information
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file DynamicsGraph.h
+ * @brief Builds a lean dynamics graph from a Robot object.
+ * @author Dan Barladeanu
+ */
+
+#include "gtdynamics/dynamics/ChainDynamicsGraph.h"
+
+namespace gtdynamics {
+
+std::vector<std::vector<JointSharedPtr>> ChainDynamicsGraph::getChainJoints(const Robot& robot) {
+  std::vector<JointSharedPtr> FR(3), FL(3), RR(3), RL(3);
+
+  int loc = 0;
+  for (auto&& joint : robot.joints()) {
+    if (joint->name().find("lower") != std::string::npos) {
+      loc = 2;
+    }
+    if (joint->name().find("upper") != std::string::npos) {
+      loc = 1;
+    }
+    if (joint->name().find("hip") != std::string::npos) {
+      loc = 0;
+    }
+    if (joint->name().find("FR") != std::string::npos) {
+      FR[loc] = joint;
+    }
+    if (joint->name().find("FL") != std::string::npos) {
+      FL[loc] = joint;
+    }
+    if (joint->name().find("RR") != std::string::npos) {
+      RR[loc] = joint;
+    }
+    if (joint->name().find("RL") != std::string::npos) {
+      RL[loc] = joint;
+    }
+  }
+
+  std::vector<std::vector<JointSharedPtr>> chain_joints{FL, FR, RL, RR};
+
+  return chain_joints;
+}
+
+Chain BuildChain(std::vector<JointSharedPtr>& joints) {
+  auto j0 = joints[0];
+  auto j1 = joints[1];
+  auto j2 = joints[2];
+
+  // Calculate all relevant relative poses.
+  Pose3 M_T_H = j0->pMc();
+  Pose3 M_H_T = M_T_H.inverse();
+  Pose3 M_H_U = j1->pMc();
+  Pose3 M_U_H = M_H_U.inverse();
+  Pose3 M_U_L = j2->pMc();
+  Pose3 M_L_U = M_U_L.inverse();
+  Pose3 M_T_L = M_T_H * M_H_U * M_U_L;
+  Pose3 M_L_T = M_T_L.inverse();
+
+  // Create chains
+  Chain chain1(M_T_H, j0->cScrewAxis());
+  Chain chain2(M_H_U, j1->cScrewAxis());
+  Chain chain3(M_U_L, j2->cScrewAxis());
+
+  std::vector<Chain> chains{chain1, chain2, chain3};
+
+  Chain composed = Chain::compose(chains);
+
+  return composed;
+}
+
+std::vector<Chain> ChainDynamicsGraph::getComposedChains(
+    std::vector<std::vector<JointSharedPtr>>& chain_joints) {
+  Chain composed_fr, composed_fl, composed_rr, composed_rl;
+
+  composed_fl = BuildChain(chain_joints[0]);
+  composed_fr = BuildChain(chain_joints[1]);
+  composed_rl = BuildChain(chain_joints[2]);
+  composed_rr = BuildChain(chain_joints[3]);
+
+  std::vector<Chain> composed_chains{composed_fl, composed_fr, composed_rl,
+                                     composed_rr};
```

Comment:
Any reason you're not looping here? It would make the code agnostic to quadrupeds, bipeds, hexapods, etc.